Non blocking smart led

robamu opened the pull request.

tschundler commented:
I finally got around to some testing tonight. It only seems to only work with 2 of 6 ws28xx variants.

On my esp32, I measure about 25us between pixels and about 55us total. 

From some research like https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/amp/ and the various datasheets, depending on the chip, the last time might be ~6us, ~50us, or ~250us

on the other 4, only one LED lights. **So if this is accepted, it needs to be accepted with the caveat that it depends heavily on which LED are used** Incidentally among those not working are what I'm pretty sure are genuine WorldSemi LEDs. 

But... exciting news, I am able to drive two sets of the type that work in parallel using a `join`!
![PXL_20241017_072807003 NIGHT RAW-01 COVER](https://github.com/user-attachments/assets/82a11f9d-a32b-4ff0-bdb2-6260c38b80bd)
(don't mind the bad grounding causing ghost signals)

But... here, the time between pixels is closer to 34us

I should go back to exploring double buffering in RMT async writes. Supposedly there's an ability now to force code into RAM to run faster?

tschundler commented:
one more very odd thing

with a buffer size of 65, I can only send 62 pixels of data. Otherwise I get `BufferSizeExceeded`. To send 64, I need to make a 67 pixel buffer.


```rust

#[main]
async fn main(spawner: Spawner) {
    let peripherals = esp_hal::init(esp_hal::Config::default());
    esp_println::logger::init_logger_from_env();

    esp_println::println!("Init!");

    let timg0 = TimerGroup::new(peripherals.TIMG0);
    esp_hal_embassy::init(timg0.timer0);

    let io = Io::new(peripherals.GPIO, peripherals.IO_MUX);

    let led_pin = io.pins.gpio18;
    let l2_pin = io.pins.gpio23;

    let rmt = Rmt::new_async(peripherals.RMT, 80.MHz()).unwrap();

    let rmt_buffer = smartLedBuffer!(67);
    let rmt2_buffer = smartLedBuffer!(67);

    let mut l1 = asynch::SmartLedAdapterAsync::new(rmt.channel0, led_pin, rmt_buffer);
    let mut l2 = asynch::SmartLedAdapterAsync::new(rmt.channel1, l2_pin, rmt2_buffer);

    esp_println::println!("Start!");

    let mut ofs = 0;
    loop {
        let data = {
            let mut i = 0;
            [(); 64].map(|_| {
                i += 1;
                hsv2rgb(Hsv {
                    hue: (((i * 4) + ofs + 60) % 255) as u8,
                    sat: 240,
                    val: 10,
                })
            })
        };

        let f1 = l1.write(data[0..64].iter().cloned());
        let f2 = l2.write(data[0..64].iter().cloned());
        //let (o1, o2) = embassy_futures::join::join(f1, f2).await;
        let o1 = f1.await;
        let o2 = f2.await;
        o1.unwrap_or_else(|e| {
            esp_println::println!("Error1: {:?}", e);
        });
        o2.unwrap_or_else(|e| {
            esp_println::println!("Error2: {:?}", e);
        });

        ofs = (ofs + 1) % 255;
        esp_println::print!(".");
        Timer::after(Duration::from_millis(10)).await;
    }
}

```

changing the above to `SmartLedBuffer!(64);` will return errors when running.

lulingar commented on esp-hal-smartled/Cargo.toml:
```diff
@@ -14,18 +14,18 @@ targets  = ["riscv32imac-unknown-none-elf"]
 defmt             = { version = "0.3.8", optional = true }
 document-features = "0.2.10"
-esp-hal           = "0.20.0"
+esp-hal           = ">=0.20.0, <=0.21.0"
```
Any reason why to exclude esp-hal 0.21.1?

robamu replied:
hmm, no. I just wonder whether <=0.21 or <=0.21.* would be cleaner/better (or even does the job)

robamu replied:
https://docs.rs/semver/latest/semver/enum.Op.html The first way works :)

lulingar replied:
Heh I see! It seems a bit misleading before a good Samaritan points you to the SemVer docs 😉

lulingar replied:
Oh it seems something is missing, this bit me:
```
error: failed to select a version for `esp-hal`.
    ... required by package `esp-hal-smartled v0.13.0`
    ... which satisfies dependency `esp-hal-smartled = "^0.13.0"` of package `test-xtensa-nostd v0.1.0 (/home/being/code/practice/rust/test-xtensa-nostd)`
versions that meet the requirements `^0.20.0` are: 0.20.1

the package `esp-hal` links to the native library `esp-hal`, but it conflicts with a previous package which links to `esp-hal` as well:
package `esp-hal v0.21.1`
    ... which satisfies dependency `esp-hal = "^0.21.1"` of package 
```

robamu replied:
I just fixed the dependency specifier to also allow higher versions.. Can you try 0.14? I also bumped the minor version of the crate itself.

lulingar replied:
It worked fine! I had to set in `Cargo.toml`:
```
esp-hal-smartled = { git = "https://github.com/robamu/esp-hal-community.git", rev = "3562306", features = ["esp32s3"] }
```

robamu commented:
In the second problem: the async API requires a different and slightly larger buffer size because each transfer must be 0 terminated. I have not created an async macro to declare that buffer yet, but i provided a const function which calculates the required buffer. 

It might be a reasonable idea to also add an async example?

horho77 commented:
> It might be a reasonable idea to also add an async example?
async example sounds great

robamu commented:
I can try to add an example which is similar to the existing one, except it uses the async API.

lulingar commented:
> I can try to add an example which is similar to the existing one, except it uses the async API.

Here's mine:
```
#![no_std]
#![no_main]
#![feature(type_alias_impl_trait)]

use core::{ptr::addr_of_mut, str::{self, FromStr}};

use alloc::string::String;

extern crate alloc;
use esp_alloc as _;

use esp_backtrace as _;
use esp_hal::{
    cpu_control::{CpuControl, Stack as CPUStack},
    gpio::{AnyPin, Input, Io, Level, Output, Pull},
    prelude::*,
    rmt::Rmt,
    timer::{
        timg::TimerGroup, AnyTimer
    },
    Async,
};

use esp_hal_embassy::Executor;
use embassy_executor::Spawner;
use embassy_time::{Delay, Duration, Instant, Timer};

use esp_println::println;

use esp_hal_smartled::{
    smart_led_buffer,
    asynch::SmartLedAdapterAsync,
};
use smart_leds::{
    brightness, gamma,
    hsv::{hsv2rgb, Hsv},
};


#[main]
async fn main(spawner: Spawner) -> ! {
    esp_println::logger::init_logger_from_env();

    let peripherals = esp_hal::init({
        let mut config = esp_hal::Config::default();
        config.cpu_clock = CpuClock::Clock240MHz;
        config
    });

    let timg0 = TimerGroup::new(peripherals.TIMG0);
    let timer1: AnyTimer = timg0.timer1.into();
    esp_hal_embassy::init(timer1);

    let mut cpu_control = CpuControl::new(peripherals.CPU_CTRL);
    let io = Io::new(peripherals.GPIO, peripherals.IO_MUX);

    let led_pin = io.pins.gpio48;
    let freq = 80.MHz();
    let rmt = Rmt::new_async(peripherals.RMT, freq).unwrap();
    let rmt_buffer = smart_led_buffer!(1);
    let led = SmartLedAdapterAsync::new(rmt.channel0, led_pin, rmt_buffer);

    let mut color = Hsv {
        hue: 0,
        sat: 255,
        val: 255,
    };
    let mut data;

    loop {
        for hue in 0..=255 {
            color.hue = hue;
            // Convert from the HSV color space (where we can easily transition from one
            // color to the other) to the RGB color space that we can then send to the LED
            data = [hsv2rgb(color)];
            // When sending to the LED, we do a gamma correction first (see smart_leds
            // documentation for details) and then limit the brightness to 10 out of 255 so
            // that the output it's not too bright.
            if let Err(e) = led.write(brightness(gamma(data.iter().cloned()), 20)).await {
                log::error!("Driving LED: {:?}", e);
            }
        }
    }
}
```

robamu commented:
What do you think about removing the asynch module and simply keeping the `*Async` variants in the same module?

Be-ing commented:
smart-leds-trait 0.3.1 has been released with the SmartLedsWriteAsync trait.

robamu commented:
I need to rebase this anyway.. I can look whether I can already integrate this.

robamu commented:
I now implemented the `SmartLedsWriteAsync` for the Async structure

robamu commented:
I also added an asynchronous example now. Haven´t tested it on hardware yet though. It is really similar to the hello_rgb example.

robamu commented on esp-hal-smartled/src/lib.rs:
```diff
@@ -205,3 +229,98 @@ where
+/// Function to calculate the required RMT buffer size for a given number of LEDs when using
+/// the asynchronous API. This buffer size is calculated for the asynchronous API provided by the
+/// [SmartLedsAdapterAsync]. [buffer_size] should be used for the synchronous API.
+pub const fn buffer_size_async(num_leds: usize) -> usize {
```
yes, if there is no submodule, it needs a distinct name.

robamu commented on esp-hal-smartled/src/lib.rs:
```diff
@@ -205,3 +229,98 @@ where
+
+/// Adapter taking an RMT channel and a specific pin and providing RGB LED
+/// interaction functionality.
+pub struct SmartLedsAdapterAsync<Tx, const BUFFER_SIZE: usize> {
```
Yes, it is a bit of duplication. That's why I tried to extract as much common logic as possible into free functions or re-usable functions. I think the subtle differences of those structs made it difficult to unify them. Having two distincts structs makes it explicit on type level which one is used though. It might also be that the flexibility of having two structs makes it easier to accomodate (external) changes in the future

Be-ing replied:
You may be able to get rid of the code duplication using the [bisync crate](https://docs.rs/bisync/latest/bisync/). I just did this for the [embedded-sdmmc crate](https://github.com/rust-embedded-community/embedded-sdmmc-rs/pull/176) and I'm quite happy with how the code turned out using bisync.

robamu commented:
I am currently updating my small lightstrip project, so I hope I can test this again soon :)

kaspar030 commented:
We used this on an Ulanzi Pixel Clock with a lot of (attention gathering) success on Chemitzer Linux Tage. :slightly_smiling_face:
It seems rock solid.

bjoernQ commented:
> We used this on an Ulanzi Pixel Clock with a lot of (attention gathering) success on Chemitzer Linux Tage. 🙂 It seems rock solid.

Oh wait - that thing is using ESP32 and is hackable? (Sorry for the off-topic comment)

kaspar030 commented:
> Oh wait - that thing is using ESP32 and is hackable? (Sorry for the off-topic comment)

Yup, and exposes the JTAG-over-UART on its USB-C, so no need to open it, just `espflash` it.

bjoernQ commented:
> Ulanzi Pixel Clock

Thanks a lot! ❤️ Ordered one

BlinkyStitt commented:
This looks like exactly what I need for my current project. Anything I can help with to get this merged in?

robamu commented:
I will update this, just waiting for the update for the project files, it is kind of painful to work without them

jamessizeland commented:
hmm yeah I'll pull your branch and take a look myself

robamu commented:
i actually also wanted to test this on an esp32c6 devboard I have here, just rebased as well.

MabezDev commented:
I'll dismiss my review shortly, as I don't want to block progress here.

Just an FYI, @wisp3rwind is working on a pretty radical RMT refactor. In particular the latest PR has some interesting changes which should make supporting blocking and async smartled stuff easier: https://github.com/esp-rs/esp-hal/pull/3505/files#diff-a18a0af13c09debca19ef19cda6411ea68747fb7fd9ef38a6a3c57e4a7005fb6R702.

Hopefully that link works correctly, but if it doesn't here is the snippet I was referring to:

```rust
/// RMT Channel Creator
pub struct ChannelCreator<Dm, const CHANNEL: u8, Cap>
where
    Dm: crate::DriverMode,
{
```

Channel creator will be generic over the mode (Blocking/Async), which should eventually reduce some boiler plate here.

jamessizeland commented:
can you sync this branch again? Its missing the .cargo changes

jamessizeland commented:
I've added the .cargo/config.toml files and run this with `cargo esp32c3 -p esp-hal-smartled --example hello_rgb_async` and it works nicely for me!  I'll try it on my c6 board too now.

jamessizeland commented:
yep working.  I think if you just merge in the latest changes it'll work.

jamessizeland reviewed (approved):
works on my devices.  Just needs main merging in first

wisp3rwind commented:
> I'll dismiss my review shortly, as I don't want to block progress here.
> 
> Just an FYI, @wisp3rwind is working on a pretty radical RMT refactor. In particular the latest PR has some interesting changes which should make supporting blocking and async smartled stuff easier: https://github.com/esp-rs/esp-hal/pull/3505/files#diff-a18a0af13c09debca19ef19cda6411ea68747fb7fd9ef38a6a3c57e4a7005fb6R702.
> 
> Hopefully that link works correctly, but if it doesn't here is the snippet I was referring to:
> 
> ```rust
> /// RMT Channel Creator
> pub struct ChannelCreator<Dm, const CHANNEL: u8, Cap>
> where
>     Dm: crate::DriverMode,
> {
> ```
> 
> Channel creator will be generic over the mode (Blocking/Async), which should eventually reduce some boiler plate here.

FWIW, completely over-engineering smart LED driving is how I got there.

What I have locally is a completely interrupt-driven async RMT writer, which uses an `Encoder` type (similar to the IDF driver) to directly generate pulse codes from RGB(WW) arrays in the ISR and write it to the RMT channel RAM, without any intermediate buffers that would contain pulse codes.

I'll attempt to upstream all of this, but it's going to take a while: In particular, the ISR case raises a bunch of issues regarding safety/soundness of the code (because the future needs to take ownership of the buffer/Encoder for this to be sound on drop of leak of the future), and I don't think how I designed it is ideal yet. I also tried an async implementation where the buffer is refilled on poll of the `RmtTxFuture`, but that didn't seem to be fast enough and led to glitches.

robamu commented:
Rebased on upstream again.

@wisp3rwind @MabezDev  That sounds really cool. I have a question: I actually had to re-configure a pin as input to avoid phantom current through the data pin when my smart led strip was off. maybe the esp channel structure can keep a type erased flex/any pin instance in the future so we can enable stuff like temporarily re-configuring as input, and later re-configuring a pin as a RMT pin?

@jamessizeland  thanks a lot for testing. I am still wondering what kind of issue I am looking at here.. I cleaned the repo multiple times as well.. which rust version are you using?

robamu commented:
and it just disappeared :D
